sympy/polys/matrices/domainmatrix.py

```diff
@@ -30,6 +30,27 @@ def __init__(self, rows, shape, domain):
         self.shape = shape
         self.domain = domain
 
+    def __getitem__(self, key):
+        i, j = key
+        m, n = self.shape
+        i, m = getindex(i, m)
+        j, n = getindex(j, n)
+
+        if i >= m:
+            raise IndexError("row index out of range")
+        if j >= n:
+            raise IndexError("column index out of range")
+
+        if isinstance(key[0], slice) or isinstance(key[1], slice):
+            return self.from_rep(self.rep.extract_slice(slice(i, m), slice(j, n)))
+        else:
+            sdm = self.rep.extract_slice(slice(i, i+1), slice(j, j+1))
+            if sdm and sdm[0]:
+                return DomainScalar(sdm[0][0], self.domain)
+            else:
+                return DomainScalar(0, self.domain)
```

Comment:
SDM and DDM should have a `getitem` method for getting a single item.

Reply:
This should be `DomainScalar(self.domain.zero, self.domain)`

Reply:
Does this means we don't need to handle slicing cases for `DDM` and `SDM`. 
If yes, should the return type of the single element be `DomainScalar`?
Thanks.

Reply:
@oscarbenjamin, `SDM`, and `DDM` already have some kind of method for accessing the elements.
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: dm=DDM([[1, 2, 3], [4, 5, 6], [7, 8, 9]], (3, 3), ZZ)

In [3]: dm[0]
Out[3]: [1, 2, 3]

In [4]: dm[0][1]
Out[4]: 2

In [5]: sdm = SDM({0:{0:ZZ(1)}}, (10, 10), ZZ)

In [6]: sdm
Out[6]: {0: {0: 1}}

In [7]: sdm[0]
Out[7]: {0: 1}

In [8]: sdm[0][0]
Out[8]: 1
```

There is no benefit in introducing `getitem` for them.

However, if we like to introduce `__getitem__`, it raises a lot of error for other test cases as they use single index accessing for both the classes.

Kindly share your opinion.
Thanks.

Reply:
That won't always work:
```python
In [5]: from sympy.polys.domainmatrix import DomainMatrix

In [6]: dM = DomainMatrix.zeros((2, 2), ZZ)

In [7]: dM.rep
Out[7]: {}

In [8]: type(dM.rep)
Out[8]: sympy.polys.matrices.sdm.SDM

In [9]: dM.rep[0][0]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-9-8f5c4417425f> in <module>
----> 1 dM.rep[0][0]

KeyError: 0
```
We should not add `__getitem__` because the dict `__getitem__` is more efficient. However `getitem` can be used by `DomainMatrix` so that it gets a consistent interface from both `SDM` and `DDM`.

Reply:
I am confused about the `getitem` method.

Does something like the below is fine for `SDM`:
On adding in `SDM`:
```
def getitem(self, key1, key2):
        i, j = key1, key2
        m, n = self.shape

        if i >= m:
            raise IndexError("row index out of range")
        if j >= n:
            raise IndexError("column index out of range")

        return self.extract_slice(slice(i, i+1), slice(j, j+1))
```
We get
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: sdm = SDM({0:{0:ZZ(1)}}, (10, 10), ZZ)

In [3]: sdm.getitem(0, 0)
Out[3]: 1

In [4]: sdm.getitem(1, 1)
Out[4]: 0
```

After the latest commit, we don't have to worry about how to access a single element in `DomainMatrix` as both `DDM.extract_slice `and `SDM.extract_slice `will handle it.

@oscarbenjamin, Kindly share your opinion.
Thanks.

Reply:
Have you done any comparison of timings for the different ways of implementing `__getitem__`?

I expect that making a specific `getitem` method would be faster than using `extract_slice`.

Reply:
I am not sure what max scale should we expect for the SDM:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: sdm = SDM({0:{0:ZZ(1)}}, (100000000000000000000, 100000000000000000000), ZZ)

In [3]: %time ok = sdm.getitem(9999999999999999999, 9999999999999999999)
Wall time: 0 ns
```

Reply:
On implementing our own `__getitem__` for SDM:

```
def __getitem__(self, key):
        from .domainmatrix import getindex
        i, j = key
        m, n = self.shape
        i, _m = getindex(i, m)
        j, _n = getindex(j, n)
        
        if i >= m:
            raise IndexError("row index out of range")
        if j >= n:
            raise IndexError("column index out of range")
        
        return self.extract_slice(slice(i, _m-1), slice(j, _n-1))
```
Timings are:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: sdm = SDM({0:{0:ZZ(1)}}, (100000000000000000000, 100000000000000000000), ZZ)

In [3]: %time ok = sdm[9999999999999999999, 9999999999999999999]
Wall time: 0 ns
```

Reply:
You might need to use `%timeit`. It doesn't actually take 0 time so that's probably just the resolution of the clock on your computer. (It obviously does not take zero time)

Absolute times are also not interesting. What matters is a comparison between two different versions of the code so we can say which is faster or slower. You need a timing that isn't just zero for that though.

Here is an alternative version that you can time against:
```python
def __getitem__(self, key):
    i, j = key
    try:
        return self[i][j]
    except KeyError:
        return self.domain.zero
```

Reply:
Actually that won't work because this is changing the `__getitem__` method of `SDM` I think. We should not define a `__getitem__` method for `SDM`. What I showed above should be the `SDM.getitem` method which will be called for `DomainMatrix.__getitem__` method.

Reply:
> You might need to use `%timeit`. It doesn't actually take 0 time so that's probably just the resolution of the clock on your computer. (It obviously does not take zero time)
> 
Indeed, absolute times are irrelevant for comparison. Sorry about that.

> Actually that won't work because this is changing the `__getitem__` method of `SDM` I think. We should not define a `__getitem__` method for `SDM`. What I showed above should be the `SDM.getitem` method which will be called for `DomainMatrix.__getitem__` method.

The problem with using this as `getitem` will be that it won't cover cases as slicing, so we need to use `extract_slice` in this `getitem`. But this will complicate, as the only purpose of `getitem` would be to handle the accessing single items and return them as DomainScalar, which the extract_slice already handles on its own.

Maybe I am missing the core reason for introducing `SDM.getitem`.
Kindly share your opinion.
Thanks.

Reply:
> the only purpose of `getitem` would be to handle the accessing single items

Yes, exactly. However it's important to understand that accessing a single item is an *extremely* common thing to do and in fact it is common to do it in the innermost loops of performance sensitive algorithms. That's why it needs to be as fast as possible.

The reason I am suggesting to have a `getitem` method is because I expect that it can implemented more simply so that it is faster than going through `extract_slice`. We need timings to test this though.

What you need to understand is that the reason `DomainMatrix` is so much faster than `Matrix` is because I have carefully optimised all of its fundamental low-level operations. If we don't worry about performance then we will just end up with something slow like `Matrix`.

Reply:
> What you need to understand is that the reason `DomainMatrix` is so much faster than `Matrix` is because I have carefully optimised all of its fundamental low-level operations. If we don't worry about performance then we will just end up with something slow like `Matrix`.

Sure, now I can feel how crucial DomainMatrix is. Will carefully add the `getitem`.
Thanks.